vault/login_mfa.go

```diff
@@ -1880,25 +1881,53 @@ func formatUsername(format string, alias *identity.Alias, entity *identity.Entit
 	return username
 }
 
-func (c *Core) validateDuo(ctx context.Context, creds []string, mConfig *mfa.Config, username, reqConnectionRemoteAddr string) error {
-	duoConfig := mConfig.GetDuoConfig()
-	if duoConfig == nil {
-		return fmt.Errorf("failed to get Duo configuration for method %q", mConfig.Name)
-	}
+type MFAFactor struct {
+	passcode string
+}
+
+func parseMfaFactors(creds []string) (*MFAFactor, error) {
+	mfaFactor := &MFAFactor{}
+	var err *multierror.Error
 
-	passcode := ""
 	for _, cred := range creds {
-		if strings.HasPrefix(cred, "passcode") {
+		switch {
+		case cred == "": // for the case of push notification
+			continue
+		case strings.HasPrefix(cred, "passcode="):
 			splits := strings.SplitN(cred, "=", 2)
-			if len(splits) != 2 {
-				return fmt.Errorf("invalid credential %q", cred)
-			}
-			if splits[0] == "passcode" {
-				passcode = splits[1]
-			}
+			mfaFactor.passcode = splits[1]
+		case strings.Contains(cred, "="):
+			err = multierror.Append(err, fmt.Errorf("found an invalid MFA cred: %v", cred))
+		default:
+			// a non-empty cred that does not match the above
+			// means it is a passcode
+			mfaFactor.passcode = cred
+		}
+
+		// currently, we only support passcode. Although we probably gain
+		// very little improvement in performance, it still makes sense
+		// to break the for loop
+		if mfaFactor.passcode != "" {
+			break
 		}
 	}
 
+	// don't return an error unless all entries were invalid
```

Comment:
Would it make sense to provide a warning to denote which entries were invalid despite finding a passcode?

Reply:
So, the place that this code is being executed expects either success or failure.  It seems that we deliberately don't want to provide more information about what has happened. For example, if validation fails, we just return permission denied and don't include any specific reason. I believe that is to prevent users to start guessing until they reach a good MFA passcode factor. 
In addition to include such warnings, it means that I would need to change several function signature in the stack which I think is beyond the scope of this project.